### PR TITLE
Google Event wrapper

### PR DESCRIPTION
**Pull request by jtkpiotr**
@pbzymek @adamzima @lkudyba @quirk0o @mhib @michal-konarski

### CODE REVIEW

File: app/models/conference_room.rb [hunk 1]
```diff
@@ -25,4 +25,10 @@ class ConferenceRoom < ApplicationRecord
   scope :with_expired_channel, -> { left_outer_joins(:channel).where(channels: {id: Channel.expired}) }
 
   scope :without_active_channel, -> { without_channel.or with_expired_channel }
+
+  def self.find_or_raise(conference_room_id)
+    conference_room = ConferenceRoom.find_by(id: conference_room_id)
+    raise Exceptions::EventInvalidRoom, "Undefined conference room: #{conference_room_id}" if conference_room.nil?
+    conference_room
```
Comment (quirk0o) [inline app/models/conference_room.rb:33]: Why not .find?

File: app/models/conference_room.rb [hunk 2]
```diff
@@ -25,4 +25,10 @@ class ConferenceRoom < ApplicationRecord
   scope :with_expired_channel, -> { left_outer_joins(:channel).where(channels: {id: Channel.expired}) }
 
   scope :without_active_channel, -> { without_channel.or with_expired_channel }
+
+  def self.find_or_raise(conference_room_id)
+    conference_room = ConferenceRoom.find_by(id: conference_room_id)
+    raise Exceptions::EventInvalidRoom, "Undefined conference room: #{conference_room_id}" if conference_room.nil?
```
Comment (mjk-old) [inline app/models/conference_room.rb]: You may find `raise (...) unless conference_room` more convenient.

File: app/models/google_calendar/event_creator.rb [hunk 3]
```diff
@@ -2,44 +2,29 @@ module GoogleCalendar
...
+      raise_error_if_occupied(event_wrapper)
+      google_event = event_wrapper.as_google_event
+      calendar_service.insert_event('primary', google_event, send_notifications: true)
     end
 
-    def raise_error_if_occupied(event_data)
-      events = events_in_span(event_data[:attendees].first,
-                              event_data[:start][:date_time], event_data[:end][:date_time])
+    def raise_error_if_occupied(wrapper)
+      events = events_in_span(wrapper.conference_room, wrapper.start_time.date_time, wrapper.end_time.date_time)
```
Comment (mjk-old) [inline app/models/google_calendar/event_creator.rb]: Wouldn't it be better to pass `wrapper` directly?

File: app/models/google_calendar/event_wrapper/builder.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+          end: google_event.end,
+          creator: google_event.creator,
+          conference_room_id: conference_room.id,
+          attendees: parse_attendees,
+          all_day: lasts_all_day?
+        }
+      end
+
+      def parse_attendees
+        return if google_event.attendees.nil?
```
Comment (mjk-old) [inline app/models/google_calendar/event_wrapper/builder.rb]: `unless?`

File: app/models/google_calendar/event_finder.rb [hunk 5]
```diff
@@ -65,6 +61,13 @@ def event_declined?(event)
       event.attendees.find(&:self).response_status == GOOGLE_EVENT_DECLINED_RESPONSE
     end
 
+    def build_event_data(google_event, conference_room)
+      wrapper_builder = GoogleCalendar::EventWrapper::Builder.new(google_event, conference_room)
```
Comment (mjk-old) [inline app/models/google_calendar/event_finder.rb]: Extract this logic to separate method, e.g. `rounded_event_wrapper`.

File: app/models/google_calendar/event_wrapper/event.rb [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+      def build_creator(params)
+        params.key?(:creator) ? params[:creator] : new_creator(params[:creator_email])
+      end
+
+      def new_creator(creator_email)
+        Google::Apis::CalendarV3::Event::Creator.new(email: creator_email)
+      end
+
+      def initialize_times(**params)
+        return unless params.key?(:start) || params.key?(:end)
```
Comment (michal-konarski) [inline app/models/google_calendar/event_wrapper/event.rb]: Shouldn't it be rather `return unless params.key?(:start) && params.key?(:end)`? This way you can pass `nil` to `build_time` and eventually get an exception.

File: app/models/google_calendar/event_wrapper/event.rb [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      def google_location
+        conference_room.nil? ? '' : conference_room.title
+      end
+
+      def google_attendees
+        attendees + [room_attendee, creator_attendee].compact
+      end
+
+      def room_attendee
+        return nil unless conference_room
```
Comment (mjk-old) [inline app/models/google_calendar/event_wrapper/event.rb]: No need to `return nil`. Bare `return` returns `nil` anyway.

File: app/models/google_calendar/event_wrapper/rounded_event.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
+module GoogleCalendar
+  module EventWrapper
+    class RoundedEvent < GoogleCalendar::EventWrapper::Event
+      attr_accessor :start_timestamp, :end_timestamp, :all_day
+
+      def initialize(**params)
+        super
+        @all_day = params.key?(:all_day) ? params[:all_day] : false
```
Comment (mjk-old) [inline app/models/google_calendar/event_wrapper/rounded_event.rb]: `@all_day = params[:all_day] || false`

File: spec/models/conference_room_spec.rb [hunk 9]
```diff
@@ -63,4 +63,22 @@
       end
     end
   end
+
+  describe '.find_or_raise' do
+    context 'given invalid conference_room_id' do
+      let(:invalid_conference_room_id) { 0 }
+      it 'raises EventInvalidRoom' do
+        expect { described_class.find_or_raise(invalid_conference_room_id) }.
```
Comment (mjk-old) [inline spec/models/conference_room_spec.rb]: Put that into `subject`

File: spec/models/google_calendar/event_wrapper/event_spec.rb [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    end
+
+
+  end
+
+  describe '#valid?' do
+    context 'given start_time and end_time in params' do
+      let(:params) { {start: {date_time: '2016-01-01'}, end: {date_time: '2016-01-02'}} }
+      it 'returns true' do
+        expect(described_class.new(params).valid?).to eq true
```
Comment (mjk-old) [inline spec/models/google_calendar/event_wrapper/event_spec.rb]:
```
let(:event) { described_class.new(params) }
subject { event.valid? }

(...)

is_expected.to be(true)
```

This applies to other part of this and other files as well :p

File: spec/models/google_calendar/event_wrapper/rounded_event_spec.rb [hunk 11]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
+
+describe GoogleCalendar::EventWrapper::RoundedEvent do
```
Comment (mjk-old) [inline spec/models/google_calendar/event_wrapper/rounded_event_spec.rb:3]: This one deserves a good refactor :p 

You can start with extracting `event`, `start`, `end`, `params` lets to describe block and put only things that differ into subcontexts.

File: spec/models/google_calendar/event_wrapper/builder_spec.rb [hunk 12]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe GoogleCalendar::EventWrapper::Builder do
+
+end
```
Comment (mjk-old) [inline spec/models/google_calendar/event_wrapper/builder_spec.rb]: Newline.

File: app/models/google_calendar/event_wrapper/event.rb [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+      def build_time(time_data)
+        time_data.is_a?(Google::Apis::CalendarV3::EventDateTime) ? time_data : new_datetime(time_data)
+      end
+
+      def new_datetime(time_data)
+        Google::Apis::CalendarV3::EventDateTime.new(date: time_data[:date], date_time: time_data[:date_time])
+      end
+
+      def google_location
+        conference_room.nil? ? '' : conference_room.title
```
Comment (mjk-old) [inline app/models/google_calendar/event_wrapper/event.rb]: `conference_room.try(:title) || ''`

However, is it really possible to have `nil` `conference_room`?

File: app/models/google_calendar/event_wrapper/event.rb [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+      end
+
+      def initialize_times(**params)
+        return unless params.key?(:start) || params.key?(:end)
+        @start_time = build_time(params[:start])
+        @end_time = build_time(params[:end])
+      end
+
+      def build_time(time_data)
+        time_data.is_a?(Google::Apis::CalendarV3::EventDateTime) ? time_data : new_datetime(time_data)
```
Comment (mjk-old) [inline app/models/google_calendar/event_wrapper/event.rb]: `is_a?` smells bad in Ruby. How it is possible that we can have two such different objects here?

File: app/models/google_calendar/event_wrapper/event.rb [hunk 15]
```diff
@@ -0,0 +1,110 @@
+module GoogleCalendar
```
Comment (mjk-old) [inline app/models/google_calendar/event_wrapper/event.rb:1]: This class looks like it has too many responsibilities. E.g. it shouldn't do nil checks on every parameter it works with. It should be done in some other place, maybe in EventWrapper builder?

File: spec/models/google_calendar/event_wrapper/rounded_event_spec.rb (hunk 11 above)
Comment (mjk-old) [reply spec/models/google_calendar/event_wrapper/rounded_event_spec.rb:3]: Now it's better 👍

**Comment by mjk-old** [comment]
🆗

**Comment by michal-konarski** [comment]
🆗

File: app/models/conference_room.rb [hunk 16]
```diff
@@ -25,4 +25,10 @@ class ConferenceRoom < ApplicationRecord
   scope :with_expired_channel, -> { left_outer_joins(:channel).where(channels: {id: Channel.expired}) }
 
   scope :without_active_channel, -> { without_channel.or with_expired_channel }
+
+  def self.find_or_raise(conference_room_id)
```
Comment (quirk0o) [inline app/models/conference_room.rb]: Why not #find ?

File: app/models/conference_room.rb [hunk 17]
```diff
@@ -26,4 +26,10 @@ class ConferenceRoom < ApplicationRecord
...
```
Comment (pbzymek) [inline app/models/conference_room.rb]: use Rails convention:
```
def self.find!
```

File: app/models/google_calendar/event_finder.rb [hunk 18]
```diff
@@ -72,6 +74,16 @@ def event_declined?(event)
       event.attendees.find(&:self).response_status == GOOGLE_EVENT_DECLINED_RESPONSE
     end
 
+    def build_event_data(google_event, conference_room)
```
Comment (pbzymek) [inline app/models/google_calendar/event_finder.rb:77]:
```
def build_event_wrapper(google_event, conference_room)
```

**Pull request merged**